rfcs/20200113-tf-data-service.md

```diff
@@ -0,0 +1,591 @@
+provide this guarantee.
+
+## Design Proposal
+
+The tf.data service is a master-worker system which iterates through datasets,
+producing outputs to be consumed by accelerators. The service is comprised of a
+few components:
+
+*   User-facing Python API for interacting with the tf.data service.
+*   Dataset splitting API for determining how to split up datasets for parallel
+    processing.
+*   Master and worker gRPC services.
+
+### Architecture
+
+The tf.data service is comprised of master and worker gRPC services which could
+be run in a couple of different configurations:
+
+#### Glossary
+
+**Master**: The single master coordinating the tf.data service.
+
+**Worker**: A tf.data service worker which performs dataset processing and
+provides dataset elements to consumers over RPC.
+
+**Consumer**: A machine which consumes data from the tf.data service. The
+consumer may be attached to a GPU or TPU, or use data for on-CPU training.
+
+#### Separate Cluster Architecture
+
+Each server is run on a separate host from the TensorFlow cluster. This
+configuration gives users a way to provide horizontally scaling CPU for
+processing their input pipelines and quickly feeding data to accelerators.
+
+#### Embedded Cluster Architecture
+
+without needing to provision additional compute resources. and gives all the
+benefits of the tf.data service except for horizontal scaling.
+
+#### Hybrid Architecture
+
+Users could run tf.data workers embedded in their TensorFlow cluster, and also
+run additional tf.data workers (and potentially the tf.data master) outside the
+cluster. This allows for horizontal worker scaling, while still leveraging the
+compute resources of the TensorFlow cluster for input processing.
+
+### User-facing Python API
+
+This API is how users will interact with the tf.data service from their Python
+code.
+
+```python
+def tf.data.experimental.service.distribute(address):
+  """Marks that a dataset should be processed by the tf.data service.
+
+  ds = ... # dataset to distribute
+  ds = ds.apply(tf.data.experimental.service.distribute(address))
+
+  Args:
+    address: The address of the tf.data service master.
+
+  Returns:
+    A function that can be passed to `dataset.apply()`.
+  """
+
+def tf.data.experimental.service.create_iteration(
+    dataset, num_consumers=1, num_tasks=None, deterministic=False):
+  """Begins distributed iteration over a dataset.
+
+  It is expected that the dataset contains at least one `.distribute(address)`
+  transformation, otherwise this method will print a warning and do nothing.
+
+  `create_iteration` will first register the dataset with the tf.data service
+  if it isn't already registered. It will then request the creation of
+  `num_consumers` dataset iterators which divide the dataset `num_consumers`
+  ways. The returned object can be used to read from one of the
+  iterators using
+  `tf.data.experimental.service.make_iterator(ds, obj, consumer_index)`.
```

Comment:
Why not the usual python control flow we use in regular tf.data?

Reply:
In regular tf.data there is only one way to iterate over a dataset, so it's natural to make `Dataset` an `Iterable`. But with tf.data service we iterate using `N` (non-independent) consumer streams which dynamically partition a single iteration's worth of data. We need a way for the consumers to identify to the tf.data service that they all want to read from the same iteration.

Reply:
Added content explaining why we don't support usual python control flow